dimod/generators/qap.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2024 D-Wave Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from itertools import product
+
+import numpy as np
+
+from dimod.binary_quadratic_model import BinaryQuadraticModel
+from dimod.constrained import ConstrainedQuadraticModel
+
+__all__ = ['quadratic_assignment']
+
+def quadratic_assignment(distance_matrix: np.typing.ArrayLike,
+                         flow_matrix: np.typing.ArrayLike,
+                         ) -> ConstrainedQuadraticModel:
+    """Generates a constrained quadratic model encoding a quadratic assignment problem.
+
+    Given distance and flow matrices, generates a :class:`~dimod.ConstrainedQuadraticModel` 
+    for the corresponding quadratic assignment problem. 
+
+    Args:
+        distances: numpy array containing the distance between location i and location j
+
+        flows: numpy array containing the flow between facility i and facility j
+
+    Returns:
+        The constrained quadratic model encoding the quadratic assignment problem. Variables are
+        denoted as ``x_{i}_{j}`` where ``x_{i}_{j} == 1`` means that facility ``i`` is
+        placed in location ``j``.
+
+    """
+
```

Comment:
Because we're accepting `ArrayLike`, we should add something like
```python
    distance_matrix = np.atleast_2d(np.asarray(distance_matrix))
    flow_matrix = np.atleast_2d(np.asarray(flow_matrix))
```
so that we also accept `list`s etc.